File: 542. 01 Matrix - Medium/solution.cpp
static auto _ = [] () { ios_base::sync_with_stdio(false); cin.tie(nullptr); return 0; } ();
class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& m) {
        if (m.empty() || m[0].empty()) {
            return m;
        }
        
        const int N = m.size();
        const int M = m[0].size();
        
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < M; ++j) {
                if (m[i][j]) {
                    m[i][j] = 100000;
                    
                    if (i) {
                        m[i][j] = min(m[i][j], m[i-1][j] + 1);
                    }
                    if (j) {
                        m[i][j] = min(m[i][j], m[i][j-1] + 1);
                    }
                }
            }
        }
        
        for (int i = N-1; i >= 0; --i) {
            for (int j = M-1; j >= 0; --j) {
                if (m[i][j]) {
                    if (i < N-1) {
                        m[i][j] = min(m[i][j], m[i+1][j] + 1);
                    }
                    if (j < M-1) {
                        m[i][j] = min(m[i][j], m[i][j+1] + 1);
                    }
                }
            }
        }
                
        return m;
    }
};

File: 542. 01 Matrix - Medium/solution.py
class Solution:
    def updateMatrix(self, matrix: List[List[int]]) -> List[List[int]]:

        m, n = len(matrix), len(matrix and matrix[0])
        for i in range(m):
            for j in range(n):
                if matrix[i][j] != 0:
                    matrix[i][j] = float("inf")
                    if i > 0:
                        matrix[i][j] = min(matrix[i][j], matrix[i-1][j] + 1)
                    if j > 0:
                        matrix[i][j] = min(matrix[i][j], matrix[i][j-1] + 1)

        for i in range(m-1, -1, -1):
            for j in range(n-1, -1, -1):
                if matrix[i][j] != 0:
                    if i + 1 < m:
                        matrix[i][j] = min(matrix[i][j], matrix[i+1][j] + 1)
                    if j + 1 < n:
                        matrix[i][j] = min(matrix[i][j], matrix[i][j+1] + 1)

        return matrix